src/json.ts

```diff
@@ -1,12 +1,39 @@
 import deepEqual from 'fast-deep-equal';
+import {
+  array,
+  boolean,
+  Infer,
+  is,
+  lazy,
+  literal,
+  nullable,
+  number,
+  object,
+  omit,
+  optional,
+  record,
+  string,
+  Struct,
+  union,
+  unknown,
+} from 'superstruct';
 import {
   calculateNumberSize,
   calculateStringSize,
-  hasProperty,
   isPlainObject,
   JsonSize,
 } from './misc';
 
+// Note: This struct references itself, so TypeScript cannot infer the type.
```

Comment:
This seems like a non-insignificant drawback to using Superstruct. It means that for each kind of thing we want we always have to have a struct and a type to go along with that struct. Ideally, those two things should match (or the type should be a subset of the type that the struct represents). I checked over your types and it does seem like that's the case, however, what if we need to support some other JSON-RPC feature in the future? Someone might not really know how this is set up. Is there any way to synchronize the two things? Like, what happens if the Json type, for instance, is just plain wrong?

Reply:
> This seems like a non-insignificant drawback to using Superstruct. It means that for each kind of thing we want we always have to have a struct and a type to go along with that struct. [...] what if we need to support some other JSON-RPC feature in the future?

I'm not 100% sure what you mean. The `Json` struct is a special case here, because it can be recursive. In this case TypeScript is not able to infer the type. That's a limitation in TypeScript, not in Superstruct.

For all of the other structs, the proper type is inferred from the struct.

> Is there any way to synchronize the two things? Like, what happens if the Json type, for instance, is just plain wrong?

We could possibly do some kind of testing to make sure that the types match, but I personally don't think it's necessary. These things can be caught during reviews, especially since recursive structs probably aren't so common in JSON-RPC.

Reply:
Okay, gotcha. After thinking more about this, I see that the problems I have with this really do have to do more with TypeScript limitations than what Superstruct is exactly doing. That's kind of lame, but nothing much we can do about it now. I guess we'll just have to be careful in reviews as you say.